src/blocks/weather.rs

```diff
@@ -215,20 +284,43 @@ pub async fn run(config: &Config, api: &CommonApi) -> Result<()> {
             None
         };
 
-        let fetch = || provider.get_weather(location);
+        let fetch = || provider.get_weather(location, need_forecast(&format, &format_alt));
         let data = fetch.retry(&ExponentialBuilder::default()).await?;
-
-        let mut widget = Widget::new().with_format(format.clone());
-        widget.set_values(data.into_values());
-        api.set_widget(widget).await?;
-
-        select! {
-            _ = sleep(config.interval.0) => (),
-            _ = api.wait_for_update_request() => ()
+        let data_values = data.into_values();
+
+        loop {
+            let mut widget = Widget::new().with_format(format.clone());
+            widget.set_values(data_values.clone());
+            api.set_widget(widget).await?;
+
+            select! {
+                _ = sleep(config.interval.0) => break,
+                _ = api.wait_for_update_request() => break,
+                Some(action) = actions.recv() => match action.as_ref() {
+                        "toggle_format" => {
+                            if let Some(ref mut format_alt) = format_alt {
+                                std::mem::swap(format_alt, &mut format);
+                            }
+                        }
+                        _ => (),
+                    }
+            }
         }
     }
 }
 
+fn need_forecast(format: &Format, format_alt: &Option<Format>) -> bool {
+    fn has_forecast_key(format: &Format) -> bool {
+        format.contains_key("temp_forecast")
+            || format.contains_key("apparent_forecast")
+            || format.contains_key("humidity_forecast")
+            || format.contains_key("wind_forecast")
+            || format.contains_key("wind_kmh_forecast")
+            || format.contains_key("direction_forecast")
+    }
+    has_forecast_key(format) || format_alt.as_ref().map_or(false, has_forecast_key)
```

Comment:
By the way, we now have `Option::is_some_and(f)` which is equivalent to `Option::map_or(false, f)` which may be more readable.